### ISSUE TITLE
`QuarkusTestBeforeClassCallback` Quarkus JUnit 5 callback fires for every test method

### ISSUE THREAD

**Issue by msavy** [quarkusio/quarkus#35692]
### Describe the bug

According to the class contract of `QuarkusTestBeforeClassCallback`:

> Can be implemented by classes that shall be called once for each `@QuarkusTest` class before any test methods are executed.
> The implementing class has to be deployed as service provider on the class path.

This should be equivalent to `@BeforeAll` in JUnit 5, IIUC from the code and various discussion threads. 

### Expected behavior

For a given test class, `QuarkusTestBeforeClassCallback.beforeClass` should be called only once.

### Actual behavior

`QuarkusTestBeforeClassCallback.beforeClass` is called before every test method.

### How to Reproduce?

* Reproducer: https://github.com/msavy/quarkus-reproducers (main branch)
* Run tests (`CallbackTest.java`)

```java
@QuarkusTest
public class CallbackTest {

    @Test
    @Order(1)
    public void a() {
        System.out.println("Test a");
    }

    @Test
    @Order(2)
    public void b() {
        System.out.println("Test b");
    }

    @Test
    @Order(3)
    public void c() {
        System.out.println("Test c");
    }
}

```

```java
package org.acme.callback;

/**
 * @author Marc Savy {@literal <marc@blackparrotlabs.io>}
 */
public class ShouldOnlyBeCalledOnce implements io.quarkus.test.junit.callback.QuarkusTestBeforeClassCallback {

    @Override
    public void beforeClass(Class<?> testClass) {
        System.err.println("Called " + ShouldOnlyBeCalledOnce.class.getCanonicalName());
    }
}

```

Serviceloader metadata omitted, you can see it in it.

Actual output:

```
2023-09-01 23:46:20,119 INFO  [io.quarkus] (main) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.3.1) started in 1.825s. Listening on: http://localhost:8081
2023-09-01 23:46:20,146 INFO  [io.quarkus] (main) Profile test activated. 
2023-09-01 23:46:20,146 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]
Called org.acme.callback.ShouldOnlyBeCalledOnce
Test a
Called org.acme.callback.ShouldOnlyBeCalledOnce
Test b
Called org.acme.callback.ShouldOnlyBeCalledOnce
Test c
2023-09-01 23:46:20,451 INFO  [io.quarkus] (main) code-with-quarkus stopped in 0.022s

Process finished with exit code 0
```

### Output of `uname -a` or `ver`

Darwin mmbp.local 22.6.0 Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "17.0.8.1" 2023-08-24

### Quarkus version or git rev

3.3.1 (same on 3.3.0)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

❯ mvn --version Running `/Users/msavy/oss/apiman/apiman/mvnw`... Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537) Maven home: /Users/msavy/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4 Java version: 17.0.8.1, vendor: Eclipse Adoptium, runtime: /Users/msavy/.sdkman/candidates/java/17.0.8.1-tem Default locale: en_GB, platform encoding: UTF-8 OS name: "mac os x", version: "13.5.1", arch: "x86_64", family: "mac"

### Additional information

https://github.com/quarkusio/quarkus/blob/main/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java#L721-L722

https://github.com/quarkusio/quarkus/blob/main/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java#L728

**Comment by mkouba** [comment]
I can confirm that it's a bug. The `QuarkusTestBeforeClassCallback` callback is executed during the invocation of `InvocationInterceptor#interceptTestClassConstructor()` and this JUnit callback is invoked for each test method by default, i.e. with `@TestInstance(Lifecycle.PER_TEST)`. It does work correctly if `@TestInstance(Lifecycle.PER_CLASS)` is used.

**Comment by geoand** [comment]
> and this JUnit callback is invoked for each test method by default, i.e. with @TestInstance(Lifecycle.PER_TEST)

Ah... I guess we will have to rely on our state in that case...

**Comment by mkouba** [comment]
> > and this JUnit callback is invoked for each test method by default, i.e. with @testinstance(Lifecycle.PER_TEST)
> 
> Ah... I guess we will have to rely on our state in that case...

Can't we just move the invocation of `QuarkusTestBeforeClassCallback` in the `QuarkusTestExtension.beforeAll()` method?

**Comment by geoand** [comment]
We can certainly try 😃

**Comment by msavy** [comment]
> > > and this JUnit callback is invoked for each test method by default, i.e. with @testinstance(Lifecycle.PER_TEST)
> > 
> > 
> > Ah... I guess we will have to rely on our state in that case...
> 
> Can't we just move the invocation of `QuarkusTestBeforeClassCallback` in the `QuarkusTestExtension.beforeAll()` method?

I've already tried that locally and using the beforeAll JUnit5 callback works fine (I can submit a PR).

The only question in my mind is whether "before class callback" is actually a different use case to "before all"? 

Clearly, the existing documentation doesn't match today's behaviour and suggests it is meant to be `@BeforeAll` — and none of the existing Quarkus callbacks provide the necessary `@BeforeAll` functionality. But, given the way that JUnit works (i.e. it always reinstantiates the class before each method call), would there be any reason we need both? 

I suspect any other use cases are already covered by the existing callbacks (before each, after constructor, etc), but I just wanted to put that out there as a point to consider.

**Comment by geoand** [comment]
Unfortunately my answer at this point is that I do not remember, so feel free to open the PR and assuming CI passes, I'll have a look in due time

**Comment by msavy** [comment]
Thanks. In the meanwhile I'm using a `Set<Class<?>>` check to emulate `@BeforeAll`, so I am not blocked.
